hail/src/main/scala/is/hail/expr/ir/agg/CollectAggregator.scala

```diff
@@ -0,0 +1,102 @@
+package is.hail.expr.ir.agg
+
+import is.hail.asm4s._
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.expr.ir._
+import is.hail.expr.types.physical._
+import is.hail.utils._
+import is.hail.io.{CodecSpec, InputBuffer, OutputBuffer}
+
+/*
+ * Collect() aggregator
+ */
+
+class CollectAggregator(val elemType: PType) extends StagedAggregator {
+
+  val resultType = PArray(elemType.setRequired(false), required = true)
+
+  class State(val mb: EmitMethodBuilder) extends AggregatorState {
+    override def regionSize: Int = Region.REGULAR
+
+    /*** State ***/
+
+    val r = mb.newField[Region]
+    val region = r.load
+    val bll = new StagedBlockLinkedList(elemType, mb)
+
+    /*** Initialization ***/
+
+    val isLoaded: Code[Boolean] = region.isValid
+
+    def createState: Code[Unit] =
+      region.isNull.orEmpty(Code(
+        r := Code.newInstance[Region, Int](regionSize),
+        region.invalidate()))
+
+    def newState: Code[Unit] = Code(
+      region.getNewRegion(regionSize),
+      bll.init(region))
+
+    /*** load, store, copy ***/
+
+    val storageType = bll.storageType
+
+    def load(regionLoader: Code[Region] => Code[Unit], src: Code[Long]): Code[Unit] =
+      Code(regionLoader(region), bll.load(src))
+
+    def store(regionStorer: Code[Region] => Code[Unit], dst: Code[Long]): Code[Unit] =
+      region.isValid.orEmpty(Code(regionStorer(region), bll.store(dst), region.invalidate()))
+
+    def copyFrom(src: Code[Long]): Code[Unit] = {
+      val totalCount = region.loadInt(
+        storageType.fieldOffset(src,
+          storageType.fieldIdx("totalCount")))
+
+      totalCount.cne(0).orEmpty(
+        Code._fatal("unimplemented: copyFrom() when src is not empty"))
+    }
+
+    /*** serialize, deserialize ***/
+
+    lazy val aoff = mb.newField[Long]
+
+    def serialize(codec: CodecSpec): Code[OutputBuffer] => Code[Unit] = {
+      val enc = codec.buildEmitEncoderF[Long](resultType, resultType, mb.fb)
+      ({ ob: Code[OutputBuffer] =>
+        Code(
+          // TODO: write directly instead of allocating an array
```

Comment:
this would be super great to do before this gets merged---I think you can just encode block by block. Leads to a slightly less efficient decoded representation since you're not combining blocks, though.